### DISCUSSION TITLE
Extend retry statusCode in order to work for 404.

### DISCUSSION THREAD

**Discussion by y0x** [sindresorhus/got#1813] [Q&A] [answered]
I’m making HTTP requests to an API where I can expect an HTTP code 404 under certain conditions. I’d like to use Got's internal retry functionality for rerunning the request until the 404 error is gone (which will happen; I just don't know if it takes 1 minute or 30 minutes).

From the documentation I know that HTTP code 404 is not a supported statusCode for the built-in retry functionality and therefore I cannot perform any action within the beforeRetry hook of Got.

I’m extending a Got instance to allow some presets for the API I’m calling. For now I have not found a clean way to extend the existing retry statusCodes with 404.

```js
const x= got.extend({
    prefixUrl: ".." + "..",
    mutableDefaults: true,
    responseType: 'json',
    //retry: { statusCode: got.defaults.options.retry.statusCode.push(404) }, //does not work | not clean
    https: { rejectUnauthorized: false },
    headers: {
        'PRIVATE-TOKEN': "..",
        Accept: 'application/json',
    },
    hooks: {
        beforeError: [
            (error) => {
                const { response } = error;
                console.log(response);
                /*
                Another idea: if I cannot extend retry statusCodes then I´d like to somehow force a retry from here
                if (response.statusCode === 404 && response.path.endsWith('/export/download')) {
                    console.log('FORCE A RETRY AS THE DOWNLOAD MIGHT NOT BE READY YET');
                }
                */

                if (response && response.body) {
                    error.name = 'x';
                    error.message = `${response.body.message !== undefined
                            ? response.body.message
                            : response.body.error
                        } (${response.statusCode})`;
                }

                return error;
            },
        ],
    },
});
```

1. How can I extend the HTTP statusCodes that allow running a retry?
2. If this is not possible, see my comment in the code. Is it somehow possible to force a retry manually by just using Got?

I´m aware that I could just catch the 404 and then do the retry manually, just asking for a way of doing it internal.

Thanks for the awesome lib, love it!

#### Checklist

- [x] I have read the documentation.

**Comment by sindresorhus** [comment, marked as answer]
> //retry: { statusCode: got.defaults.options.retry.statusCode.push(404) }, //does not work | not clean

```js
retry: { statusCodes: [...got.defaults.options.retry.statusCodes, 404] },
```

**Reply by y0x** [reply]
This will result in an error:

```
retry: { statusCode: [...got.defaults.options.retry.statusCode, 404] },
                                                        ^

TypeError: got.defaults.options.retry.statusCode is not iterable
    at Object.<anonymous> (index.js:7:57)
    at Module._compile (internal/modules/cjs/loader.js:1015:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1035:10)
    at Module.load (internal/modules/cjs/loader.js:879:32)
    at Function.Module._load (internal/modules/cjs/loader.js:724:14)
    at Module.require (internal/modules/cjs/loader.js:903:19)
    at require (internal/modules/cjs/helpers.js:74:18)
    at Object.<anonymous> (index.js:1:26)
    at Module._compile (internal/modules/cjs/loader.js:1015:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1035:10)
```

**Reply by titanism** [reply]
This is due to a typo, `statusCode` -> `statusCodes`.  See the option at <https://github.com/sindresorhus/got/blob/main/documentation/7-retry.md#statuscodes>.

@sindresorhus actually meant to write this:

```js
retry: { statusCodes: [...got.defaults.options.retry.statusCodes, 404] },
```

**Comment by szmarczak** [comment]
Duplicate of #1780

**Comment by titanism** [comment]
@szmarczak This is not necessarily a duplicate, as the usage is a bit different.  The reason why @sindresorhus answer could not be accepted and the reason why @y0x originally had an issue was due to a typo.  They had singular usage of `statusCode` vs. plural usage `statusCodes`.  See the comment at <https://github.com/sindresorhus/got/discussions/1813#discussioncomment-3249169>.  Instead you could edit @sindresorhus answer and then mark that as the actual answer instead.

**Reply by szmarczak** [reply]
Thank you! I didn't even notice this 😅
